### ISSUE TITLE
sql: primary key and unique constraints not visible in DBInspect tool

### ISSUE THREAD

**Issue by rytaft** [cockroachdb/cockroach#23103]
I downloaded the latest DBInspect JAR (v1.05) from this site: http://www.dbinspect.com. Thanks to the changes in #22753, I can now use this tool to view the data in CockroachDB by setting the database type to PostgreSQL. It works pretty well: I can see the data and columns, as well as fk constraints (parent/child), but I cannot see primary key or unique constraints. We should make it possible to see these as well.

**Comment by nvanbenschoten** [comment]
Thanks for catching this @rytaft! I was able to get DBInspect up and running and got to the point that you got to. Sure, enough, the tool wasn't correctly showing primary keys or unique constraints.

I then tried to see what the tool showed when connected to Postgres. Strangely, I actually wasn't able to get it to see any schemas in postgres at all. I tried running PG locally and in Docker, and tried both PG 10 and 9.2. Nothing seemed to work.

I went back to Cockroach and looked at the introspective queries that DBInspect issues to populate the information that is missing. The queries it sends are:
```
SELECT NULL::TEXT AS pktable_cat, pkn.nspname AS pktable_schem, pkc.relname AS pktable_name, pka.attname AS pkcolumn_name, NULL::TEXT AS fktable_cat, fkn.nspname AS fktable_schem, fkc.relname AS fktable_name, fka.attname AS fkcolumn_name, pos.n AS key_seq, CASE con.confupdtype WHEN 'c' THEN 0 WHEN 'n' THEN 2 WHEN 'd' THEN 4 WHEN 'r' THEN 1 WHEN 'a' THEN 3 ELSE NULL END AS update_rule, CASE con.confdeltype WHEN 'c' THEN 0 WHEN 'n' THEN 2 WHEN 'd' THEN 4 WHEN 'r' THEN 1 WHEN 'a' THEN 3 ELSE NULL END AS delete_rule, con.conname AS fk_name, pkic.relname AS pk_name, CASE WHEN con.condeferrable AND con.condeferred THEN 5 WHEN con.condeferrable THEN 6 ELSE 7 END AS deferrability FROM pg_catalog.pg_namespace AS pkn, pg_catalog.pg_class AS pkc, pg_catalog.pg_attribute AS pka, pg_catalog.pg_namespace AS fkn, pg_catalog.pg_class AS fkc, pg_catalog.pg_attribute AS fka, pg_catalog.pg_constraint AS con, generate_series(1, 32) AS pos (n), pg_catalog.pg_depend AS dep, pg_catalog.pg_class AS pkic WHERE (((((((((((((((pkn.oid = pkc.relnamespace) AND (pkc.oid = pka.attrelid)) AND (pka.attnum = con.confkey[pos.n])) AND (con.confrelid = pkc.oid)) AND (fkn.oid = fkc.relnamespace)) AND (fkc.oid = fka.attrelid)) AND (fka.attnum = con.conkey[pos.n])) AND (con.conrelid = fkc.oid)) AND (con.contype = 'f')) AND (con.oid = dep.objid)) AND (pkic.oid = dep.refobjid)) AND (pkic.relkind = 'i')) AND (dep.classid = 'pg_constraint'::REGCLASS::OID)) AND (dep.refclassid = 'pg_class'::REGCLASS::OID)) AND (pkn.nspname = 'public')) AND (pkc.relname = 'customer') ORDER BY fkn.nspname, fkc.relname, pos.n;
```
and
```
SELECT NULL::TEXT AS pktable_cat, pkn.nspname AS pktable_schem, pkc.relname AS pktable_name, pka.attname AS pkcolumn_name, NULL::TEXT AS fktable_cat, fkn.nspname AS fktable_schem, fkc.relname AS fktable_name, fka.attname AS fkcolumn_name, pos.n AS key_seq, CASE con.confupdtype WHEN 'c' THEN 0 WHEN 'n' THEN 2 WHEN 'd' THEN 4 WHEN 'r' THEN 1 WHEN 'a' THEN 3 ELSE NULL END AS update_rule, CASE con.confdeltype WHEN 'c' THEN 0 WHEN 'n' THEN 2 WHEN 'd' THEN 4 WHEN 'r' THEN 1 WHEN 'a' THEN 3 ELSE NULL END AS delete_rule, con.conname AS fk_name, pkic.relname AS pk_name, CASE WHEN con.condeferrable AND con.condeferred THEN 5 WHEN con.condeferrable THEN 6 ELSE 7 END AS deferrability FROM pg_catalog.pg_namespace AS pkn, pg_catalog.pg_class AS pkc, pg_catalog.pg_attribute AS pka, pg_catalog.pg_namespace AS fkn, pg_catalog.pg_class AS fkc, pg_catalog.pg_attribute AS fka, pg_catalog.pg_constraint AS con, generate_series(1, 32) AS pos (n), pg_catalog.pg_depend AS dep, pg_catalog.pg_class AS pkic WHERE (((((((((((((((pkn.oid = pkc.relnamespace) AND (pkc.oid = pka.attrelid)) AND (pka.attnum = con.confkey[pos.n])) AND (con.confrelid = pkc.oid)) AND (fkn.oid = fkc.relnamespace)) AND (fkc.oid = fka.attrelid)) AND (fka.attnum = con.conkey[pos.n])) AND (con.conrelid = fkc.oid)) AND (con.contype = 'f')) AND (con.oid = dep.objid)) AND (pkic.oid = dep.refobjid)) AND (pkic.relkind = 'i')) AND (dep.classid = 'pg_constraint'::REGCLASS::OID)) AND (dep.refclassid = 'pg_class'::REGCLASS::OID)) AND (pkn.nspname = 'public')) AND (pkc.relname = 'customer') ORDER BY fkn.nspname, fkc.relname, pos.n;
```

When I run the first query manually, I get the following:
```
+-------------+---------------+--------------+---------------+-------------+---------------+--------------+---------------+---------+-------------+-------------+--------------------------+---------+---------------+
| pktable_cat | pktable_schem | pktable_name | pkcolumn_name | fktable_cat | fktable_schem | fktable_name | fkcolumn_name | key_seq | update_rule | delete_rule |         fk_name          | pk_name | deferrability |
+-------------+---------------+--------------+---------------+-------------+---------------+--------------+---------------+---------+-------------+-------------+--------------------------+---------+---------------+
| NULL        | public        | customer     | c_w_id        | NULL        | public        | history      | h_c_w_id      |       1 |           3 |           3 | fk_h_c_w_id_ref_customer | primary |             7 |
| NULL        | public        | customer     | c_d_id        | NULL        | public        | history      | h_c_d_id      |       2 |           3 |           3 | fk_h_c_w_id_ref_customer | primary |             7 |
| NULL        | public        | customer     | c_id          | NULL        | public        | history      | h_c_id        |       3 |           3 |           3 | fk_h_c_w_id_ref_customer | primary |             7 |
| NULL        | public        | customer     | c_w_id        | NULL        | public        | order        | o_w_id        |       1 |           3 |           3 | fk_o_w_id_ref_customer   | primary |             7 |
| NULL        | public        | customer     | c_d_id        | NULL        | public        | order        | o_d_id        |       2 |           3 |           3 | fk_o_w_id_ref_customer   | primary |             7 |
| NULL        | public        | customer     | c_id          | NULL        | public        | order        | o_c_id        |       3 |           3 |           3 | fk_o_w_id_ref_customer   | primary |             7 |
+-------------+---------------+--------------+---------------+-------------+---------------+--------------+---------------+---------+-------------+-------------+--------------------------+---------+---------------+
```

The second gives:

```
+-------------+---------------+--------------+---------------+-------------+---------------+--------------+---------------+---------+-------------+-------------+------------------------+---------+---------------+
| pktable_cat | pktable_schem | pktable_name | pkcolumn_name | fktable_cat | fktable_schem | fktable_name | fkcolumn_name | key_seq | update_rule | delete_rule |        fk_name         | pk_name | deferrability |
+-------------+---------------+--------------+---------------+-------------+---------------+--------------+---------------+---------+-------------+-------------+------------------------+---------+---------------+
| NULL        | public        | district     | d_w_id        | NULL        | public        | customer     | c_w_id        |       1 |           3 |           3 | fk_c_w_id_ref_district | primary |             7 |
| NULL        | public        | district     | d_id          | NULL        | public        | customer     | c_d_id        |       2 |           3 |           3 | fk_c_w_id_ref_district | primary |             7 |
+-------------+---------------+--------------+---------------+-------------+---------------+--------------+---------------+---------+-------------+-------------+------------------------+---------+---------------+
```

This first result set does include the primary key. Additionally, running the same two queries against Postgres with the same schema gives the exact same results (thanks @knz!). I don't know what else to do here, and suspect that this may actually be an issue with the tool itself. Out of curiousity, what made you choose DBInspect? Have you used it before? Is it something you suspect people will want to use against Cockroach?

Closing this issue for now, but feel free to re-open if I missed something.

**Comment by rytaft** [comment]
Thanks for investigating, @nvanbenschoten! I ran this experiment on a whim as part of QA for #21870. I had read in the [RFC](https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/20180219_pg_virtual_namespacing.md) that one of the goals for the proposed changes was to "unblock proper user experience of CockroachDB for users of (graphical or non-graphical) DB inspection tools", so I figured I'd try to find such a tool. I did a google search for "db inspection tool", and DBInspect was the first search result. So I have no idea if other users want to use this particular tool, but I thought this issue might be representative of other tools. 

Anyway, as you mentioned, it sounds like this may be an issue with DBInspect, not with CockroachDB. If a customer finds a similar issue we can investigate further, but I doubt it's worth spending more time on this now.
